include/geopackage/SQLite.hpp

```diff
@@ -0,0 +1,273 @@
+#ifndef NGEN_GEOPACKAGE_SQLITE_H
+#define NGEN_GEOPACKAGE_SQLITE_H
+
+#include <memory>
+#include <stdexcept>
+#include <vector>
+#include <string>
+
+#include <sqlite3.h>
+
+namespace geopackage {
+
+/**
+ * Runtime error for iterations that haven't started
+ */
+const auto sqlite_get_notstarted_error = std::runtime_error(
+    "sqlite iteration is has not started, get() is not callable (call sqlite_iter::next() before)"
+);
+
+/**
+ * Runtime error for iterations that are finished
+ */
+const auto sqlite_get_done_error = std::runtime_error(
+    "sqlite iteration is done, get() is not callable"
+);
+
+/**
+ * Deleter used to provide smart pointer support for sqlite3 structs.
+ */
+struct sqlite_deleter
+{
+    void operator()(sqlite3* db) { sqlite3_close_v2(db); }
+    void operator()(sqlite3_stmt* stmt) { sqlite3_finalize(stmt); }
+};
+
+/**
+ * Smart pointer (unique) type for sqlite3 database
+ */
+using sqlite_t = std::unique_ptr<sqlite3, sqlite_deleter>;
+
+/**
+ * Smart pointer (shared) type for sqlite3 prepared statements
+ */
+using stmt_t = std::shared_ptr<sqlite3_stmt>;
+
+/**
+ * Get a runtime error based on a function and code.
+ *
+ * @param f String denoting the function where the error originated
+ * @param code sqlite3 result code
+ * @param extra additional messages to add to the end of the error
+ * @return std::runtime_error
+ */
+inline std::runtime_error sqlite_error(const std::string& f, int code, const std::string& extra = "")
```

Comment:
Is there a particular reason to inline error reporting code? Usually it should be the coldest code path

Reply:
No particular reason -- I think I was working in-header at the time, but never moved that out. In my next push I'll include a change that defines this only within the source files, since it isn't really needed outside of the `sqlite`/`sqlite_iter` impls